### PR TITLE
Fix ZMBV CompareBlock function returning a negative value for 24/32bpp types

### PR DESCRIPTION
**Pull request by weirddan455**
# Description

Regression from 6892f70c8502d9ef790059756a7b33d2faa7f880
Above commit changed a variable from int to auto
This resulted in a logical right shift instead of the expected arithmatic right shift

This bug caused the motion vector to always be (0, 0) resulting in more inefficient compression

Found this bug while I was writing a ZMBV decoder for a video player I'm working on.

The logic is kind of hard to follow but form what I can tell `CompareBlock()` is supposed to return a positive value indicating how closely the two blocks match with 0 being (mostly?) identical and a large number meaning very different.

The bug made this function always return a negative number when in 24 or 32 bit per pixel mode.  It's a templated function so it happened when `P` is `uint32_t`.  This made the logic in `AddXorFrame()` think it always found a good match on the first try so it didn't try other motion vector possibilities.

This doesn't result in any visual glitches in the output video file so it went un-noticed for a long time.  Only observable affect is a slightly larger file size.

# Release notes

Fixed a bug with video capture that resulted in slightly larger file sizes.

# Manual testing

Recorded about 9 seconds of Crystal Caves from level 1 start to the hole in the ground.  Stubbed out audio capture as that was the bulk of the file size otherwise as audio is uncompressed.

Result was 1.49MB on this branch and 1.62MB on main

For a more scientific test, I would need to rip out the encoder and compress the exact same video but this was as close as I could get it without doing a bunch of extra work.

# Checklist

_Please tick the items as you have addressed them. Don't remove items; leave the ones that are not applicable unchecked._

I have:

- [x] followed the project's [contributing guidelines](https://github.com/dosbox-staging/dosbox-staging/blob/master/CONTRIBUTING.md) and [code of conduct](https://github.com/dosbox-staging/dosbox-staging/blob/master/CODE_OF_CONDUCT.md).
- [x] performed a self-review of my code.
- [ ] commented on the particularly hard-to-understand areas of my code.
- [ ] split my work into well-defined, bisectable commits, and I [named my commits well](https://github.com/dosbox-staging/dosbox-staging/blob/main/CONTRIBUTING.md#commit-messages).
- [x] applied the appropriate labels (bug, enhancement, refactoring, documentation, etc.)
- [x] [checked](https://github.com/dosbox-staging/dosbox-staging/blob/main/scripts/compile_commits.sh) that all my commits can be built.
- [ ] confirmed that my code does not cause performance regressions (e.g., by running the Quake benchmark).
- [ ] added unit tests where applicable to prove the correctness of my code and to avoid future regressions.
- [ ] provided the release notes draft (for significant user-facing changes).
- [ ] made corresponding changes to the documentation or the website according to the [documentation guidelines](https://github.com/dosbox-staging/dosbox-staging/blob/main/DOCUMENTATION.md).
- [ ] [locally verified](https://github.com/dosbox-staging/dosbox-staging/blob/main/DOCUMENTATION.md#previewing-documentation-changes-locally) my website or documentation changes.

### CODE REVIEW

File: src/libs/zmbv/zmbv.cpp [hunk 1]
```diff
@@ -184,11 +184,11 @@ int VideoCodec::CompareBlock(const int vx, const int vy, const FrameBlock & bloc
...
 	P *pnew = reinterpret_cast<P *>(newframe) + block.start;
-	;
+
 	for (auto y = 0; y < block.dy; y++) {
 		for (auto x = 0; x < block.dx; x++) {
-			const auto test = 0 - ((pold[x] - pnew[x]) & 0x00ffffff);
-			ret -= check_cast<int>(test >> 31);
+			const int test = 0 - ((pold[x] - pnew[x]) & 0x00ffffff);
+			ret -= (test >> 31);
 		}
```
Comment (interloper98) [inline src/libs/zmbv/zmbv.cpp:192]: That's pretty subtle! Great find, @weirddan455.

Maybe the `ret = ` could be renamed to describe the number of pixels found to be different (if that's what we're counting?)

Also, can the "zero minus" and the `ret -=` be flipped over to positives, so it reads more naturally:

```c++
    int diff_count = 0;
    ...    
    for ( ...
        for ( ...
            diff_count += ((pold[x] ^ pnew[x]) & 0x00ffffff) != 0;
        }
    }

    return diff_count;
```

(if not, maybe a comment is needed to describe why we need to subtract negatives)

**Comment by interloper98** [comment]
> Found this bug while I was writing a ZMBV decoder for a video player I'm working on.

:rocket: ; right on! ZMBV has impressed me with how space-efficient it can be for game content; lots of great ideas went into it.

File: src/libs/zmbv/zmbv.cpp (hunk 1 above)
Comment (weirddan455) [reply src/libs/zmbv/zmbv.cpp:192]: Oh, yeah I guess it is just masking out the alpha channel (which is un-unsed) and then counting the number of pixels that differ.

I was wondering what that `test >> 31` meant but it's just checking the sign bit and incrementing the return value in the most obtuse way possible :sweat_smile: 

I'll stick your code in there and run a test to see if the values are identical.  If they are, yours is a lot more readable so I'll go with that.

Comment (interloper98) [reply src/libs/zmbv/zmbv.cpp:192]: > it's just checking the sign bit and incrementing the return value in the most obtuse way possible

:100: I guess some of these elite coding patterns were for performance reasons (or something!) :hurtrealbad:

Comment (weirddan455) [reply src/libs/zmbv/zmbv.cpp]: Ok, yours wins.  Added a log statement that fired if the values differed from the existing code and it was the same in both EGA (16bpp) and VGA (32bpp).

The only thing I can figure is the original author was trying to avoid a branch that the `!= 0` might cause but I took a quick look at the assembly out of curiosity and couldn't tell much difference.  The compiler was pretty smart and inlined and auto-vectorized a lot of this code with SIMD.

Changed to match your suggestion for readability :smile:

Comment (interloper98) [reply src/libs/zmbv/zmbv.cpp]: > The compiler was pretty smart and inlined and auto-vectorized a lot of this code with SIMD.

Nice!

**Review by interloper98 — APPROVED** [review]
Looking forward to even smaller .avi's after this!

File: src/libs/zmbv/zmbv.cpp (hunk 1 above)
Comment (johnnovak) [reply src/libs/zmbv/zmbv.cpp]: > > it's just checking the sign bit and incrementing the return value in the most obtuse way possible
> 
> 💯 . I guess some of these elite coding patterns were for performance reasons (or something!) :hurtrealbad:

Standard 90's game /demo coder fare -- nothing unusual to see, moving on 😎 

Plus yeah, optimising C compilers were in their infancy 20-30 years ago compared to what we have now. One way to make C code from the 80/90s fly with almost zero effort is to simply recompile it with clang/gcc instead or Watcom/Turbo C/etc. (if you are able to find your way around the DOS era nemory model bullshit in the modern compiler).

**Comment by weirddan455** [comment]
@johnnovak are you good with merging this?

**Comment by johnnovak** [comment]
> @johnnovak are you good with merging this?

Yeah, 0.82.0 final material.
